### PR TITLE
[interop] update names and add docs for the interop C++ helper macros

### PR DESCRIPTION
**Pull request by hyp**

### CODE REVIEW

**Review by zoecarver — APPROVED** [review]
🤴

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
@swift-ci please test

**Comment by haoNoQ** [comment]
Hmm, did you consider prefixing them with a bunch of __underscores, to avoid collisions with existing user code? I could totally see it both ways. Like, NS_RETURNS_RETAINED isn't prefixed with __underscores, so maybe it's fine. But given that you're building a language feature that could be useful outside of Apple's constrained ecosystem, it might make sense to keep these macros in language-feature-y style. While extremely unlikely, it's theoretically possible that even the C++ standard itself could want to reserve some of these identifiers in the future, so it could get really awkward.

**Comment by haoNoQ** [comment]
I also really like it that these names aren't alluding to Swift! These attributes tell us things about the C++ classes regardless of Swift interop, they can be useful for other applications, such as static analysis.

I still think there could be a catchy prefix for these attributes all together, so that the reader could clearly see that they're part of the same facility and didn't need to look up each one separately, in order to learn that when they're reading the code. But it looks like other compiler features (eg. https://clang.llvm.org/docs/ThreadSafetyAnalysis.html) didn't go down that route; so maybe my intuition isn't quite correct here.

**Comment by ktoso** [comment]
When we used such attrs elsewhere we prefixed with SWIFT_ I guess, so that might be a possibility?

**Comment by hyp** [comment]
> When we used such attrs elsewhere we prefixed with SWIFT_ I guess, so that might be a possibility?

For now we thought that we would rather avoid SWIFT_ prefix.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
@swift-ci please build toolchain

**Comment by hyp** [comment]
@swift-ci please test macOS platform

**Comment by hyp** [comment]
@swift-ci please build toolchain

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

File: test/Interop/Cxx/class/invalid-class-errors.swift [hunk 1]
```diff
@@ -30,15 +30,15 @@ import Test
 // CHECK: note: record 'A' is not automatically available: does not have a copy constructor or destructor. Does this type have reference semantics?
 // CHECK: struct A {
 // CHECK: ^
-// CHECK: SWIFT_REFERENCE_TYPE(<#retain#>, <#release#>)
+// CHECK: SHARED_REFERENCE(<#retain#>, <#release#>)
```
Comment (zoecarver) [inline test/Interop/Cxx/class/invalid-class-errors.swift:33]: I think these need `SWIFT_`

File: include/swift/AST/DiagnosticsClangImporter.def [hunk 2]
```diff
@@ -186,11 +186,11 @@ NOTE(projection_reference_not_imported, none, "C++ method '%0' that returns a re
...
+NOTE(mark_self_contained, none, "Mark type '%0' as 'SWIFT_SELF_CONTAINED' in C++ to "
                                  "make methods that use it available in Swift. ",
      (StringRef))
-NOTE(mark_safe_to_import, none, "Mark method '%0' as 'SAFE_TO_IMPORT' in C++ to "
-                                "make it available in Swift. ",
+NOTE(mark_safe_to_import, none, "annotate method '%0' with 'SWIFT_RETURNS_INDEPENDENT_VALUE' in C++ to "
+                                "make it available in Swift",
      (StringRef))
 
 NOTE(at_to_subscript, none, "Do you want to replace it with a call "
```
Comment (zoecarver) [inline include/swift/AST/DiagnosticsClangImporter.def:196]: Non blocking, but if it's easy, there's a reference to "import_reference" in this file that should probably be updated. Not a big deal, though.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
source compat failures are unrelated.

**Pull request merged**